test/functional/test_framework/mininode.py

```diff
@@ -492,7 +492,7 @@ def test_function():
 # P2PConnection acquires this lock whenever delivering a message to a P2PInterface.
 # This lock should be acquired in the thread running the test logic to synchronize
 # access to any data shared with the P2PInterface or P2PConnection.
-mininode_lock = threading.RLock()
+mininode_lock = threading.Lock()
```

Comment:
Actually I have a question because I'm not 100% sure how the test_runner works - does it allot a certain amount of time for each test? It looks like it would run infinitely? How would it affect Travis if 1 test is in deadlock?

Reply:
Travis is running our ci system, which calls the test_runner helper:

```sh
  DOCKER_EXEC LD_LIBRARY_PATH=$DEPENDS_DIR/$HOST/lib ${TEST_RUNNER_ENV} test/functional/test_runner.py --ci $MAKEJOBS --tmpdirprefix "${BASE_SCRATCH_DIR}/test_runner/" --ansi --combinedlogslen=4000 ${TEST_RUNNER_EXTRA} --quiet --failfast
```

If a test takes a long time, it will show up in the log as `Remaining jobs: [foo_bar.py]`.

If a test never finished, the virtual machine that the ci system runs on will eventually reboot and mark the run as failed.

Reply:
> would it be a good idea to make mininode_lock use something likelock.acquire(timeout=60) and/or if lock.locked(): raise_assertion_error

We use the `with <lock>:` syntax everywhere, which doesn't allow a `blocking` or `timeout` argument. I don't think there's a way to set a default time on a lock.

I think if you did accidentally introduce a deadlock, you'd figure it out pretty easily. If your script was hanging, I think the best way would be to implement some way to attach pdb to the running process (eg https://stackoverflow.com/a/25329467/933705 or https://stackoverflow.com/a/147114/933705) and then inspect the stack trace.

(in fact, I think this would be generally useful, so perhaps someone should implement it :slightly_smiling_face: )

Reply:
Ok this makes sense, thanks! I was worried I could accidentally stall Travis hahaha. Having some debugging help sounds useful, let me go look for a someone...